### DISCUSSION TITLE
Async Connection Pool

### DISCUSSION THREAD

**Discussion by jmillerCE** [psycopg/psycopg#258] [Q&A] [answered]
Hello,

I am looking into the async connection pool and was curious about why we wouldn't await getting a connection from the pool but would await getting a connection without the pool. Is there something I am missing?
```
async with my_pool.connection() as conn:
    await conn.execute(...)
```

vs.

```
async with await psycopg.AsyncConnection.connect() as aconn:
    async with aconn.cursor() as cur:
        await cur.execute(...)
```

**Comment by dvarrazzo** [comment, marked as answer]
I don't understand your question. What do you mean that "we wouldn't await getting a connection from the pool"?

**Reply by jmillerCE** [reply]
Sorry if I was unclear. What I meant is why do we not add await before calling `my_pool.connection()`?

I am still a bit new to python async so it might just be a misunderstanding on my part.

**Reply by dvarrazzo** [reply]
Ok I see what you mean. It is `async with await AsyncConnection.connection()` to be strange here. The reason for which this function must be called that way is that sometimes people will need a connection not tied to a context, so they may want to do:

```python
aconn = await AsyncConnection.connection(...)
# use the connection
await aconn.execute(...)
await aconn.close()
```

Other times they would like to use the connection as a context manager. Ideally they might do:

```python
async with AsyncConnection.connection(...) as aconn:  # this doesn't work
    await aconn.execute(...)
```

The latter doesn't quite work with async connections, because `AsyncConnection.connection()` returns a coroutine, which doesn't have an `__aenter__()` method, so it's not "with-able". So, as far as I can see, we cannot support both the use cases with the syntax above and we came out with `async with await`: the `await` consumes the coroutine and produces an `AsyncConnection()`, which is a context manager: it has an `__aenter__()` method so it can be passed to `async with`. The quirk is explained [in the docs](https://www.psycopg.org/psycopg3/docs/advanced/async.html#with-async-connections).

In the case of the pool, we _don't want_ to support both use cases: we want to support primarily the `async with` syntax and we don't want people to call `conn = await pool.connect()`, because it's easy to "forget" to put back the connection in the pool, which eventually would stop working. We do that by making `pool.connection()` an async context manager, not just a normal async function, and as a consequence it doesn't require the extra `await`.

**Reply by jmillerCE** [reply]
Completely makes sense. Thanks so much for taking the time to explain the process to me!
